File: src/Cpp/FingerTree/tests/rope_text_tests.cpp
#include <durable7/finger_tree/finger_tree.hpp>

#include "test_support/test_runner.hpp"

#include <algorithm>
#include <concepts>
#include <cstddef>
#include <sstream>
#include <string>
#include <string_view>
#include <vector>

namespace ft = durable7::finger_tree;
using namespace durable7::finger_tree::tests;

namespace {

static_assert(std::same_as<ft::text_rope_cursor, ft::measured_rope_cursor<char, ft::newline_measure>>);
static_assert(std::same_as<
    ft::text_rope_cursor_search_result,
    ft::measured_rope_cursor_search_result<char, ft::newline_measure>>);

[[nodiscard]] std::vector<std::string> split_lines(const std::string_view text)
{
    auto result = std::vector<std::string>{};
    auto current = std::string{};
    for (const auto value : text) {
        if (value == '\n') {
            result.push_back(std::move(current));
            current = std::string{};
        } else {
            current.push_back(value);
        }
    }

    result.push_back(std::move(current));
    return result;
}

[[nodiscard]] std::vector<std::size_t> line_starts(const std::vector<std::string>& lines)
{
    auto starts = std::vector<std::size_t>{};
    starts.reserve(lines.size());
    auto running = std::size_t{0};
    for (const auto& line : lines) {
        starts.push_back(running);
        running += line.size() + 1;
    }

    return starts;
}

void require_text_model(const std::string& text)
{
    const auto char_rope = ft::to_char_rope(text);
    const auto rope = ft::to_text_rope(text);
    const auto expected_lines = split_lines(text);
    const auto starts = line_starts(expected_lines);
    const auto newline_count = static_cast<std::size_t>(std::ranges::count(text, '\n'));

    FT_REQUIRE(ft::as_string(char_rope) == text);
    FT_REQUIRE(ft::as_string(rope) == text);
    FT_REQUIRE_EQUAL(rope.measure(), newline_count);
    FT_REQUIRE_EQUAL(ft::line_count(rope), expected_lines.size());
    FT_REQUIRE(ft::lines(rope) == expected_lines);

    for (auto line = std::size_t{0}; line < expected_lines.size(); ++line) {
        FT_REQUIRE_EQUAL(ft::line_start_offset(rope, line), starts[line]);
        FT_REQUIRE(ft::get_line(rope, line) == expected_lines[line]);
    }

    for (auto offset = std::size_t{0}; offset <= text.size(); ++offset) {
        const auto expected_line = static_cast<std::size_t>(
            std::ranges::count(std::string_view{text.data(), offset}, '\n'));
        FT_REQUIRE_EQUAL(ft::line_of_offset(rope, offset), expected_line);

        const auto position = ft::line_column_of(rope, offset);
        FT_REQUIRE_EQUAL(position.line, expected_line);
        FT_REQUIRE_EQUAL(position.column, offset - starts[expected_line]);
        FT_REQUIRE_EQUAL(ft::offset_of(rope, position.line, position.column), offset);
    }
}

void add_rope_text_tests_impl(suite& tests)
{
    tests.add("rope text string interop and line navigation match string model", [] {
        const auto samples = std::vector<std::string>{
            "",
            "no newlines here",
            "a\nb\nc",
            "a\nbb\nccc",
            "trailing\n",
            "\nleading",
            "\n\n\n",
            "line0\nline1\n\nline3\n"};

        for (const auto& text : samples) {
            require_text_model(text);
        }
    });

    tests.add("rope text editing round-trips through measured rope", [] {
        const auto rope = ft::to_text_rope("hello world");
        const auto inserted = rope.insert_range(5, std::string{",\nthere"});

        FT_REQUIRE(ft::as_string(inserted) == "hello,\nthere world");
        FT_REQUIRE_EQUAL(ft::line_count(inserted), static_cast<std::size_t>(2));
        FT_REQUIRE(ft::get_line(inserted, 0) == "hello,");
        FT_REQUIRE(ft::get_line(inserted, 1) == "there world");

        const auto removed = inserted.remove_range(5, 7);
        FT_REQUIRE(ft::as_string(removed) == "hello world");
        FT_REQUIRE_EQUAL(ft::line_count(removed), static_cast<std::size_t>(1));
    });

    tests.add("rope builder materializes char and text ropes", [] {
        auto builder = ft::rope_builder{};
        builder.append("hello").append(' ').append("world").append_line().append_line("tail");

        const auto expected = std::string{"hello world\ntail\n"};
        FT_REQUIRE_EQUAL(builder.length(), expected.size());
        FT_REQUIRE(builder.str() == expected);
        FT_REQUIRE(ft::as_string(builder.to_rope()) == expected);
        const auto text = builder.to_text_rope();
        FT_REQUIRE(ft::as_string(text) == expected);
        FT_REQUIRE_EQUAL(ft::line_count(text), static_cast<std::size_t>(3));
        FT_REQUIRE(ft::get_line(text, 1) == "tail");
        FT_REQUIRE(ft::get_line(text, 2).empty());

        builder.clear().append("reset");
        FT_REQUIRE_EQUAL(builder.length(), static_cast<std::size_t>(5));
        FT_REQUIRE(ft::as_string(builder.to_text_rope()) == "reset");
    });

    tests.add("rope text large document line lookups", [] {
        auto text = std::string{};
        for (auto index = 0; index != 10000; ++index) {
            text += "line ";
            text += std::to_string(index);
            text += '\n';
        }

        const auto rope = ft::to_text_rope(text);
        FT_REQUIRE_EQUAL(ft::line_count(rope), static_cast<std::size_t>(10001));
        FT_REQUIRE(ft::get_line(rope, 0) == "line 0");
        FT_REQUIRE(ft::get_line(rope, 3141) == "line 3141");
        FT_REQUIRE(ft::get_line(rope, 9999) == "line 9999");
        FT_REQUIRE(ft::get_line(rope, 10000).empty());

        const auto offset = ft::offset_of(rope, 3141, 2);
        FT_REQUIRE_EQUAL(rope[offset], 'n');
        const auto position = ft::line_column_of(rope, offset);
        FT_REQUIRE_EQUAL(position.line, static_cast<std::size_t>(3141));
        FT_REQUIRE_EQUAL(position.column, static_cast<std::size_t>(2));
    });

    tests.add("rope text long edit script preserves snapshots and navigation", [] {
        auto text = std::string{};
        for (auto line = 0; line != 240; ++line) {
            text += "line-";
            text += std::to_string(line);
            text += ':';
            text.push_back(static_cast<char>('a' + (line % 26)));
            text.push_back(static_cast<char>('A' + ((line * 7) % 26)));
            text.push_back(static_cast<char>('0' + (line % 10)));
            text.push_back('\n');
        }

        auto rope = ft::to_text_rope(text);
        const auto snapshot = rope;
        const auto snapshot_text = text;
        require_text_model(text);

        for (auto step = 0; step != 180; ++step) {
            if (step % 5 == 1 && !text.empty()) {
                const auto index = (static_cast<std::size_t>(step) * 53U + 17U) % text.size();
                const auto count = std::min<std::size_t>((step % 4) + 1, text.size() - index);
                rope = rope.remove_range(index, count);
                text.erase(index, count);
            } else {
                const auto index = (static_cast<std::size_t>(step) * 97U + 11U) % (text.size() + 1);
                auto inserted = std::string{};
                if (step % 5 == 0) {
                    inserted = "\nsection-" + std::to_string(step) + "\n";
                } else {
                    inserted.push_back(static_cast<char>('!' + (step % 57)));
                    inserted.push_back(static_cast<char>('0' + (step % 10)));
                }

                rope = rope.insert_range(index, inserted);
                text.insert(index, inserted);
            }

            if (step % 17 == 0) {
                FT_REQUIRE(ft::as_string(rope) == text);
                require_text_model(text);
                FT_REQUIRE(ft::as_string(snapshot) == snapshot_text);
                require_text_model(snapshot_text);
            }
        }

        FT_REQUIRE(ft::as_string(rope) == text);
        require_text_model(text);
        FT_REQUIRE(ft::as_string(snapshot) == snapshot_text);
        require_text_model(snapshot_text);
    });

    tests.add("rope text cursor preserves byte offsets line measures and exact text snapshots", [] {
        const auto text = ft::to_text_rope("a\r\nb");
        const auto start = text.get_cursor();
        FT_REQUIRE_EQUAL(start.size(), static_cast<std::size_t>(4));
        FT_REQUIRE((ft::line_column_of(start) == ft::line_column{0, 0}));
        FT_REQUIRE(start.snapshot().begin() == text.begin());

        const auto newline = text.get_cursor_by_measure([](const std::size_t count) {
            return count >= 1;
        });
        FT_REQUIRE(newline.found);
        FT_REQUIRE_EQUAL(newline.cursor.position(), static_cast<std::size_t>(2));
        FT_REQUIRE_EQUAL(*newline.cursor.try_peek_next(), '\n');
        FT_REQUIRE_EQUAL(newline.cursor.measure_before(), static_cast<std::size_t>(0));
        FT_REQUIRE_EQUAL(newline.cursor.measure_after(), static_cast<std::size_t>(1));
        FT_REQUIRE((ft::line_column_of(newline.cursor.move_next()) == ft::line_column{1, 0}));

        const auto edited = text.get_cursor(1)
            .insert_range(std::string{"x\n"})
            .replace_next('Z');
        FT_REQUIRE_EQUAL(edited.position(), static_cast<std::size_t>(3));
        FT_REQUIRE(ft::as_string(edited.snapshot()) == "ax\nZ\nb");
        FT_REQUIRE((ft::line_column_of(edited) == ft::line_column{1, 0}));
        FT_REQUIRE(ft::as_string(text) == "a\r\nb");

        const auto second_newline = edited.seek_by_measure([](const std::size_t count) {
            return count >= 2;
        });
        FT_REQUIRE(second_newline.found);
        FT_REQUIRE_EQUAL(second_newline.cursor.position(), static_cast<std::size_t>(4));

        const auto miss = edited.seek_by_measure([](const std::size_t count) {
            return count > 2;
        });
        FT_REQUIRE(!miss.found);
        FT_REQUIRE(miss.cursor.is_at_end());
        FT_REQUIRE(ft::as_string(miss.cursor.snapshot()) == "ax\nZ\nb");

        FT_REQUIRE_THROWS(std::runtime_error, edited.seek_by_measure([](const std::size_t) -> bool {
            throw std::runtime_error("text cursor predicate failure");
        }));
        FT_REQUIRE_EQUAL(edited.position(), static_cast<std::size_t>(3));
        FT_REQUIRE(ft::as_string(edited.snapshot()) == "ax\nZ\nb");
    });

    tests.add("rope text argument validation", [] {
        const auto rope = ft::to_text_rope("a\nb");
        FT_REQUIRE_THROWS(std::out_of_range, ft::line_start_offset(rope, 3));
        FT_REQUIRE_THROWS(std::out_of_range, ft::line_of_offset(rope, 4));
        FT_REQUIRE_THROWS(std::out_of_range, ft::offset_of(rope, 1, 100));
        FT_REQUIRE_THROWS(std::out_of_range, ft::get_line(rope, 3));
    });

    tests.add("rope text offset_of validates the column against the line end", [] {
        const auto rope = ft::to_text_rope("ab\ncd");

        FT_REQUIRE(ft::offset_of(rope, 0, 0) == 0);
        FT_REQUIRE(ft::offset_of(rope, 0, 2) == 2);   // the newline terminating line 0
        FT_REQUIRE(ft::offset_of(rope, 1, 2) == 5);   // end of the rope on the last line

        // Columns past the line end must throw, not silently address line 1.
        FT_REQUIRE_THROWS(std::out_of_range, ft::offset_of(rope, 0, 3));
        FT_REQUIRE_THROWS(std::out_of_range, ft::offset_of(rope, 0, 4));
        FT_REQUIRE_THROWS(std::out_of_range, ft::offset_of(rope, 1, 3));
    });
}

} // namespace

void add_rope_text_tests(suite& tests)
{
    add_rope_text_tests_impl(tests);
}

File: src/CSharp/tests/Durable7.FingerTree.Tests/RopeTextTests.cs
// Tests for the rope text.

using Durable7.FingerTree;
using Xunit;

namespace Durable7.FingerTree.Tests;

/// <summary>
/// Verifies the <see cref="RopeText"/> text conveniences: string interop, the <see cref="NewlineMeasure"/> line
/// rope, line/column navigation against a <c>string.Split('\n')</c> model, the <see cref="System.IO.TextReader"/>
/// adapter, and editing-through-the-rope round trips.
/// </summary>
public sealed class RopeTextTests
{
    /// <summary>Representative texts: empty, no newlines, interior/leading/trailing newlines, and a blank line.</summary>
    public static TheoryData<string> Texts =>
    [
        "",
        "no newlines here",
        "a\nb\nc",
        "a\nbb\nccc",
        "trailing\n",
        "\nleading",
        "\n\n\n",
        "line0\nline1\n\nline3\n",
    ];

    /// <summary>Verifies string round-trips and the newline measure.</summary>
    [Theory]
    [MemberData(nameof(Texts))]
    public void StringInterop_RoundTrips(string text)
    {
        Assert.Equal(text, text.ToCharRope().AsString());
        var rope = text.ToTextRope();
        Assert.Equal(text, rope.AsString());
        Assert.Equal(text.Count(c => c == '\n'), rope.Measure);
        Assert.Equal(text.Split('\n').Length, rope.LineCount());
    }

    /// <summary>Verifies line/column navigation matches a brute-force string model at every offset and line.</summary>
    [Theory]
    [MemberData(nameof(Texts))]
    public void LineNavigation_MatchesStringModel(string text)
    {
        var rope = text.ToTextRope();
        var lines = text.Split('\n');

        // Line start offsets from the model: cumulative line lengths plus the newlines before each line.
        var starts = new int[lines.Length];
        var running = 0;
        for (var i = 0; i < lines.Length; i++)
        {
            starts[i] = running;
            running += lines[i].Length + 1;   // +1 for the newline that follows (past the end for the last line)
        }

        Assert.Equal(lines.Length, rope.LineCount());

        for (var line = 0; line < lines.Length; line++)
        {
            Assert.Equal(starts[line], rope.LineStartOffset(line));
            Assert.Equal(lines[line], rope.GetLine(line));
        }

        for (var offset = 0; offset <= text.Length; offset++)
        {
            var expectedLine = text[..offset].Count(c => c == '\n');
            Assert.Equal(expectedLine, rope.LineOfOffset(offset));

            var (line, column) = rope.LineColumnOf(offset);
            Assert.Equal(expectedLine, line);
            Assert.Equal(offset - starts[expectedLine], column);
            Assert.Equal(offset, rope.OffsetOf(line, column));   // round-trip position <-> offset
        }

        Assert.Equal(lines, rope.Lines().ToArray());
    }

    /// <summary>
    /// Verifies Lines and LinesText validate their rope argument at the call site rather than deferring the
    /// ArgumentNullException to the first enumeration of the returned sequence.
    /// </summary>
    [Fact]
    public void Lines_ValidatesEagerly()
    {
        MeasuredRope<char, int, NewlineMeasure> rope = null!;
        Assert.Throws<ArgumentNullException>(() => rope.Lines());
        Assert.Throws<ArgumentNullException>(() => rope.LinesText());
    }

    /// <summary>
    /// Verifies OffsetOf accepts columns up to the line length (the newline / end-of-rope position)
    /// and rejects columns past the line's end rather than silently addressing a later line.
    /// </summary>
    [Fact]
    public void OffsetOf_ValidatesColumnAgainstTheLineEnd()
    {
        var rope = "alpha\nbeta\n".ToTextRope();

        Assert.Equal(0, rope.OffsetOf(0, 0));
        Assert.Equal(5, rope.OffsetOf(0, 5));      // the newline terminating line 0
        Assert.Equal(8, rope.OffsetOf(1, 2));
        Assert.Equal(10, rope.OffsetOf(1, 4));     // the newline terminating line 1
        Assert.Equal(11, rope.OffsetOf(2, 0));     // trailing empty line, end of rope

        // Offsets 6..7 are on line 1; a column past line 0's end must not reach them.
        Assert.Throws<ArgumentOutOfRangeException>("column", () => rope.OffsetOf(0, 6));
        Assert.Throws<ArgumentOutOfRangeException>("column", () => rope.OffsetOf(0, 7));
        Assert.Throws<ArgumentOutOfRangeException>("column", () => rope.OffsetOf(2, 1));
        Assert.Throws<ArgumentOutOfRangeException>("column", () => rope.OffsetOf(0, -1));
    }

    /// <summary>Verifies the TextReader adapter reads exactly the rope's characters, with working Peek.</summary>
    [Theory]
    [MemberData(nameof(Texts))]
    public void TextReader_ReadsAllCharacters(string text)
    {
        using var reader = text.ToCharRope().AsTextReader();
        Assert.Equal(text, reader.ReadToEnd());

        using var peeking = text.ToTextRope().AsTextReader();
        var built = new System.Text.StringBuilder();
        int peeked;
        while ((peeked = peeking.Peek()) != -1)
        {
            Assert.Equal(peeked, peeking.Read());   // Peek then Read return the same character
            built.Append((char)peeked);
        }

        Assert.Equal(text, built.ToString());
    }

    /// <summary>Verifies editing through the text rope (insert/remove) round-trips to the expected string.</summary>
    [Fact]
    public void Editing_RoundTripsToString()
    {
        var rope = "hello world".ToTextRope();

        var inserted = rope.InsertRange(5, ",\nthere".AsSpan());
        Assert.Equal("hello,\nthere world", inserted.AsString());
        Assert.Equal(2, inserted.LineCount());
        Assert.Equal("hello,", inserted.GetLine(0));
        Assert.Equal("there world", inserted.GetLine(1));

        var removed = inserted.RemoveRange(5, 7);   // remove ",\nthere"
        Assert.Equal("hello world", removed.AsString());
        Assert.Equal(1, removed.LineCount());
    }

    /// <summary>Verifies a large, heavily-lined document supports O(log n) line lookups correctly.</summary>
    [Fact]
    public void LargeDocument_LineLookups()
    {
        var builder = new System.Text.StringBuilder();
        for (var i = 0; i < 50_000; i++)
            builder.Append("line ").Append(i).Append('\n');
        var text = builder.ToString();
        var rope = text.ToTextRope();

        Assert.Equal(50_001, rope.LineCount());   // 50,000 newlines + 1 trailing empty line
        Assert.Equal("line 0", rope.GetLine(0));
        Assert.Equal("line 31415", rope.GetLine(31415));
        Assert.Equal("line 49999", rope.GetLine(49999));
        Assert.Equal(string.Empty, rope.GetLine(50000));

        var offset = rope.OffsetOf(31415, 2);   // the '2' would be inside "line 31415"... column 2 is 'n'
        Assert.Equal('n', rope[offset]);
        Assert.Equal((31415, 2), rope.LineColumnOf(offset));
    }
}
